File: KantaroWpf/Converters/CanzoneFlowDocumentConverter.cs
using Enzo.Music;
using Microsoft.VisualBasic;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Data;
using System.Windows.Documents;

namespace Enzo.Music.KantaroWpf.Converters;

class CanzoneFlowDocumentConverter : IValueConverter
{
    #region Proprietà per la rappresentazione grafica del documento
    public int TitoloFontSize = 14;
    public FontStyle TitoloFontStyle = FontStyles.Normal;
    public FontWeight TitoloFontWeight = FontWeights.Bold;
    public int AutoreFontSize = 12;
    public FontStyle AutoreFontStyle = FontStyles.Italic;
    public FontWeight AutoreFontWeight = FontWeights.Regular;
    #endregion
    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        if (value is null || value is not Canzone) return DependencyProperty.UnsetValue;
        if (targetType != typeof(FlowDocument))
            throw new InvalidOperationException("Il tipo di destinazione deve essere un FlowDocument");
        var strofeDict = new Dictionary<string, Strofa>();

        var doc = new FlowDocument();
        Canzone song = (Canzone)value;
        if (!string.IsNullOrWhiteSpace(song.Titolo))
        {
            var par = new Paragraph()
            {
                FontSize = TitoloFontSize,
                FontStyle = TitoloFontStyle,
                FontWeight = TitoloFontWeight
            };
            par.Inlines.Add(song.Titolo);
            doc.Blocks.Add(par);
        }
        if (!string.IsNullOrWhiteSpace(song.Autore))
        {
            var par = new Paragraph()
            {
                FontSize = AutoreFontSize,
                FontStyle = AutoreFontStyle,
                FontWeight = AutoreFontWeight
            };
            par.Inlines.Add(song.Autore);
            doc.Blocks.Add(par);
        }
        var noteTrovate = (from strofa in song.GetStrofe()
                           from p in strofa.Parti
                           where p?.Accordo is not null
                           select p.Accordo?.Scala.NotaFondamentale).ToArray();
        var noteNaturaliTrovate = (from n in noteTrovate where n.IsNaturale select n.Valore).Distinct().ToArray();
        var altreNoteTrovate = (from n in noteTrovate where !n.IsNaturale select n.Valore).Distinct().ToArray();
        var noteBemolli = new Dictionary<NotaEnum, bool>();
        foreach (var nota in altreNoteTrovate)
            noteBemolli.Add(nota, false);
        bool trovatoBemolle;
        do
        {
            trovatoBemolle = false;
            foreach (var nota in altreNoteTrovate)
            {
                Nota nuovaNota = new Nota(nota);
                var existItsNat = noteNaturaliTrovate.Contains((nuovaNota - Semitono.Value).Valore) ||
                                  noteBemolli.ContainsKey((nuovaNota - Tono.Value).Valore) &&
                                  noteBemolli[(nuovaNota - Tono.Value).Valore];
                if (!nuovaNota.IsNaturale && existItsNat && !noteBemolli[nota])
                {
                    noteBemolli[nota] = true;
                    trovatoBemolle = true;
                }
            }
        } while (trovatoBemolle);
        foreach (var strofa in song.GetStrofe())
        {
            Paragraph b = new Paragraph();
            if (!string.IsNullOrWhiteSpace(strofa.Nome) && !strofeDict.ContainsKey(strofa.Nome))
                strofeDict.Add(strofa.Nome, strofa);
            Strofa strofaDaConvertire;
            if (strofa is StrofaRipetuta)
                strofaDaConvertire = strofeDict[((StrofaRipetuta)strofa).Riferimento];
            else
                strofaDaConvertire = strofa;
            foreach (var p in strofaDaConvertire.Parti)
            {
                if (p is not null)
                {
                    string testoAccordo;
                    if (p.Accordo is null)
                    {
                        testoAccordo = string.Empty;
                    }
                    else
                    {
                        var nuovaNota = p.Accordo.Scala.NotaFondamentale;
                        var bemolle = noteBemolli.ContainsKey(nuovaNota.Valore) && noteBemolli[nuovaNota.Valore];
                        var partiConAccordiSimili = from parte in strofaDaConvertire.Parti
                                                where parte?.Accordo?.Scala?.NotaFondamentale == nuovaNota
                                                select parte;
                        if (bemolle && partiConAccordiSimili.FirstOrDefault() is not null)
                            bemolle = false;
                        testoAccordo = p.Accordo.ToString(bemolle);
                    }
                    TextBlock tb = new TextBlock();
                    tb.Inlines.Add($"{testoAccordo}\r\n{p.Testo}");
                    b.Inlines.Add(tb);
                }
                else
                {
                    b.Inlines.Add("\r\n");
                }
            }
            doc.Blocks.Add(b);
        }
        return doc;
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
        return DependencyProperty.UnsetValue;
    }
}

File: CanzoniereWPF/Converters/CanzoneFlowDocumentConverter.vb
Public Class CanzoneFlowDocumentConverter
    Implements IValueConverter

#Region " Proprietà per la rappresentazione grafica del documento "
    Public Property TitoloFontSize As Integer = 14
    Public Property TitoloFontStyle As FontStyle = FontStyles.Normal
    Public Property TitoloFontWeight As FontWeight = FontWeights.Bold
    Public Property AutoreFontSize As Integer = 12
    Public Property AutoreFontStyle As FontStyle = FontStyles.Italic
    Public Property AutoreFontWeight As FontWeight = FontWeights.Regular
#End Region

    Public Function Convert(value As Object, targetType As Type, parameter As Object, culture As Globalization.CultureInfo) As Object Implements IValueConverter.Convert
        If value Is Nothing OrElse Not (TypeOf value Is Canzone) Then Return DependencyProperty.UnsetValue
        If targetType <> GetType(FlowDocument) Then
            Throw New InvalidOperationException("Il tipo di destinazione deve essere un FlowDocument")
        End If
        Dim strofeDict As New Generic.Dictionary(Of String, Strofa)

        Dim doc As New FlowDocument
        Dim song = DirectCast(value, Canzone)
        If Not String.IsNullOrWhiteSpace(song.Titolo) Then
            Dim par As New Paragraph() With {.FontSize = TitoloFontSize, .FontStyle = TitoloFontStyle, .FontWeight = TitoloFontWeight}
            par.Inlines.Add(song.Titolo)
            doc.Blocks.Add(par)
        End If
        If Not String.IsNullOrWhiteSpace(song.Autore) Then
            Dim par As New Paragraph() With {.FontSize = AutoreFontSize, .FontStyle = AutoreFontStyle, .FontWeight = AutoreFontWeight}
            par.Inlines.Add(song.Autore)
            doc.Blocks.Add(par)
        End If
        Dim noteTrovate = (From strofa In song.GetStrofe()
                           From p In strofa.Parti
                           Where p IsNot Nothing AndAlso p.Accordo IsNot Nothing
                           Select p.Accordo.Scala.NotaFondamentale).ToArray
        Dim noteNaturaliTrovate = (From n In noteTrovate Where n.IsNaturale Select n.Valore).Distinct.ToArray
        Dim altreNoteTrovate = (From n In noteTrovate Where Not n.IsNaturale Select n.Valore).Distinct.ToArray
        Dim noteBemolli As New Generic.Dictionary(Of NotaEnum, Boolean)
        For Each nota In altreNoteTrovate
            noteBemolli.Add(nota, False)
        Next
        Dim trovatoBemolle As Boolean
        Do
            trovatoBemolle = False
            For Each nota In altreNoteTrovate
                Dim nuovaNota = New Nota(nota)
                Dim existItsNat = noteNaturaliTrovate.Contains((nuovaNota - Semitono.Value).Valore) OrElse
                                  noteBemolli.ContainsKey((nuovaNota - Tono.Value).Valore) AndAlso noteBemolli((nuovaNota - Tono.Value).Valore)
                If Not nuovaNota.IsNaturale AndAlso existItsNat AndAlso Not noteBemolli(nota) Then
                    noteBemolli(nota) = True
                    trovatoBemolle = True
                End If
            Next
        Loop While trovatoBemolle
        For Each strofa In song.GetStrofe()
            Dim b As New Paragraph()
            If Not String.IsNullOrWhiteSpace(strofa.Nome) Then
                If Not strofeDict.ContainsKey(strofa.Nome) Then strofeDict.Add(strofa.Nome, strofa)
            End If
            Dim strofaDaConvertire As Strofa
            If TypeOf strofa Is StrofaRipetuta Then
                strofaDaConvertire = strofeDict(DirectCast(strofa, StrofaRipetuta).Riferimento)
            Else
                strofaDaConvertire = strofa
            End If
            For Each p In strofaDaConvertire.Parti
                If p IsNot Nothing Then
                    Dim testoAccordo As String
                    If p.Accordo Is Nothing Then
                        testoAccordo = String.Empty
                    Else
                        Dim nuovaNota = p.Accordo.Scala.NotaFondamentale
                        Dim bemolle = noteBemolli.ContainsKey(nuovaNota.Valore) AndAlso noteBemolli(nuovaNota.Valore)
                        Dim partiConAccordiSimili = From parte In strofaDaConvertire.Parti
                                                    Where parte IsNot Nothing AndAlso
                                                    parte.Accordo IsNot Nothing AndAlso
                                                    parte.Accordo.Scala IsNot Nothing AndAlso
                                                    parte.Accordo.Scala.NotaFondamentale = nuovaNota
                                                    Select parte
                        If bemolle AndAlso partiConAccordiSimili.FirstOrDefault IsNot Nothing Then
                            bemolle = False
                        End If
                        testoAccordo = p.Accordo.ToString(bemolle)
                    End If
                    Dim tb As New TextBlock()
                    tb.Inlines.Add(testoAccordo & vbCrLf & p.Testo)
                    b.Inlines.Add(tb)
                Else
                    b.Inlines.Add(vbCrLf)
                End If
            Next
            doc.Blocks.Add(b)
        Next
        Return doc
    End Function

    Public Function ConvertBack(value As Object, targetType As Type, parameter As Object, culture As Globalization.CultureInfo) As Object Implements IValueConverter.ConvertBack
        Return DependencyProperty.UnsetValue
    End Function
End Class
